src/S3/PostObjectV4.php

```diff
@@ -0,0 +1,214 @@
+<?php
+namespace Aws\S3;
+
+use Aws\Credentials\CredentialsInterface;
+use GuzzleHttp\Psr7\Uri;
+use Aws\Signature\SignatureTrait;
+use Aws\Common\Enum\DateFormat;
+
+/**
+ * Encapsulates the logic for getting the data for an S3 object POST upload form
+ *
+ * @link http://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPOST.html
+ * @link http://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-post-example.html
+ */
+class PostObjectV4
+{
+    use SignatureTrait;
+
+    const ISO8601_BASIC = 'Ymd\THis\Z';
+
+    private $client;
+    private $bucket;
+    private $formAttributes;
+    private $formInputs = [];
+    private $jsonPolicy;
+    private $credentials;
+
+    /**
+     * Constructs the PostObject.
+     *
+     * The options array accepts the following keys:
+     * @link http://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html
+     *
+     * @param S3ClientInterface $client     Client used with the POST object
+     * @param string            $bucket     Bucket to use
+     * @param array             $options    Policy condition options
+     * @param string            $expiration Upload expiration date. By
+     *                                      default: 1 hour vaild peroid.
+     */
+    public function __construct(
+        S3ClientInterface $client,
+        $bucket,
+        array $options = [],
+        $expiration = NULL
+    ) {
+
+        date_default_timezone_set('UTC');
+        $this->client = $client;
+        $this->bucket = $bucket;
+
+        // prepare form attributes
+        $this->formAttributes = [
+            'action'  => $this->generateUri(),
+            'method'  => 'POST',
+            'enctype' => 'multipart/form-data'
+        ];
+        // setup basic policy documents
+        if (is_null($expiration)) {
+            $expiration = gmdate('Y-m-d\TH:i:s\Z', strtotime('+1 hours'));
+        }
+        $policy = [
+            'expiration' => $expiration,
+            'conditions' => [],
+        ];
+        // policy condition updates
+        $starts_with = []; // handle start-with conditions differently
+        $start_with_key = '';
+        $pool = ['x-amz-credential', 'x-amz-algorithm', 'x-amz-date'];
+        foreach ($options as $option) {
```

Comment:
All these loops and branches make a large surface area for bugs. Could this be made simpler by just taking a list of conditions, using [Boto3's `generate_presigned_post`](https://botocore.readthedocs.org/en/latest/reference/services/s3.html#S3.Client.generate_presigned_post) as an example?

Instead of taking an array of options in the constructor, take `$conditions` as an array of arrays. Some of them will be a single key => value pair like `['acl' => 'public-read']` , while others will be a three-member tuple like `['starts-with', '$Content-Type', 'image/']`.